Cmake build system

mpreisler opened the pull request:
Another moonshot project ;-) Porting the entire build system over to cmake. I am opening this pull request for discussion and jenkins builds along the way, to make sure I don't break the scripts while porting.

Advantages of cmake over our current build system:
- perfect incremental builds, rebuilding will be faster
- configuration time - enable and disable builds of things you need/don't need
- better parallel builds with correct dependencies
- out of source builds
- clean system, predictable builds! this is not really a flaw of `make` or Makefiles but more of our approach in SSG. We copy files around and do various weird hacks. In most cases there is no reason to do any of that...
- shared targets done cleanly - shared remediations, shared contributors.xml, etc... it doesn't get built for every product

mpreisler commented:
So far I can build Fedora final XCCDF using cmake. The results are almost the same with minor differences in fix script formatting and ordering. I have a feeling that OVAL will be harder to pull off.

mpreisler commented:
Note to myself: I need to rebase this and strip the combineremediations.py commit once https://github.com/OpenSCAP/scap-security-guide/pull/1382 is merged.

mpreisler commented:
I know we discussed cmake in the past and agreed but I would still like some supporting votes here from core contributors, @iankko @shawndwells @redhatrises.

iankko commented:
@mpreisler 
> Another moonshot project ;-) Porting the entire build system over to cmake. I am opening this pull request for discussion and jenkins builds along the way, to make sure I don't break the scripts while porting.

You have my wholehearted support in this effort under assumption during the rewrite we don't break the existing functionality (read like there would be less OVALs or remediations included due the rewrite). If we had some check the state (WRT to count of included OVAL checks and remediations) are identical before and after the change, I don't have objections against the rewrite (not counting the fact it would simplify a lot of currently worthlessly re-used duplicate copies of code).

Do we need to solve the issue how to deal with duplicate XCCDFs prior this can be implemented?

> Advantages of cmake over our current build system:
>     * perfect incremental builds, rebuilding will be faster

Nice.

>    *  configuration time - enable and disable builds of things you need/don't need

This could help to overcome the current "hack" with OVAL-5.11 support. Of course in the moment, SCAP-1.3 officially supports OVAL-5.11, we can probably completely flip just to OVAL-5.11 version. But till that moment we need to support both version (OVAL-5.10 as the last official one, OVAL-5.11 as the bleeding edge / future one). Right now doing that via ugly hack. Maybe cmake approach could simplify it (make it more transparent).

>     * better parallel builds with correct dependencies

Again. Full support. When building SSG on localhost, we tend to make things sequential. But in buildsystems there often is optimization to split build on multiple cores. In the past we even got hit by the issue SSG build properly on single host, but failed on host, where more cores were enforce (IIRC it was in the moment of making 0.1.29 Fedora updates, when I first needed to fix the parallel build in order to be able to deliver the packages). Thus if cmake could help with this (would help ensure we won't regress WRT to building SSG in parallel), it's only welcomed.

>     *  out of source builds

IMHO support for this feature would take for a separated PR itself. We currently often tend to presume the scripts / XSLTs are in the locations, like located in the git repository (IOW there's lot of hardcoded paths around the code). But I support the idea of OOSB since it could allow to build SSG under separate conditions / assumptions.

>      *  clean system, predictable builds! this is not really a flaw of make or Makefiles but more of our approach in SSG. We copy files around and do various weird hacks. In most cases there is no reason to do any of that...

Two cases of this come to mind directly (the OVAL-5.11 support and the "introduce new product" effort). For the former it was finished in pretty small timeframe (to have things working on both versions). I agree the implementation it's not ideal, it could be improved. The "almost no progress" clause WRT to this rewrite is there partly because it's hard to obtain engineering time for infrastructure changes (rewriting the way things work isn't such highlight like having 50+ new rules in the content). But I agree from sustainability PoV of the project, we need to maintain the infrastructure.

For the "introduce new product" part a lot has been said already. We copy things, because it's quick way how to get "complete content for the new product". Of course the negative effect of the copying being duplicate files, and not reviewed content for the very least (something that used to work on version - 2, might fail hardly on version. You won't find this sooner, than in the moment trying to apply the version - 2 content on version system. This often anyway ends up in the need to rewrite the corresponding rules completely). Sometimes interactive approach (take one rule at a time, verify it works on system of version, and apply only after verification) might be more productive.

>         * shared targets done cleanly - shared remediations, shared contributors.xml, etc... it doesn't get built for every product

IMHO the issue here being we are "learning on the fly". Meaning, we wouldn't design the concept of ```shared/``` folder at the beginning, and create the code to follow the design document. Vice versa - at some point we realised we need the ```shared/``` folder (basically it's introduction was just "effort to see if it works"). From that time point we are more and more stuff into ```shared/``` (not only OVAL, and remediations, but also templates now, and maybe even XCCDF in the future). And thus being hit by performance issues (at the beginning of the ```shared/``` folder it wasn't important the implementation is ineffective, since it was just attempt to see if code generation will be working). But from that moment we are just loading more and more stuff into ```shared/``` without optimizing the ```shared/``` tools (partly again due to the fact that ```infrastructure issues are hardly to be justified as being necessary to be performed right now```). So if cmake can improve the quality of the ```shared/``` functionality, it will be only benefitial for the SSG's expansion in the future.

redhatrises commented:
@mpreisler 

> Another moonshot project ;-) Porting the entire build system over to cmake. 

+1 

The only thing I would like to see as part of this PR outside of @iankko's comment about keeping functionality is some sort of documentation for a would-be contributor. In other words, I am a new contributor providing a new product tree, so what do I need to modify/add to get my new product to build?

radzy commented:
+1000000 on redhatrises' comment:
a getting-started guide would be very nice!

mpreisler commented:
@redhatrises @radzy Agreed, the new system should be much simpler than the current arcane black magic needed to start a new SSG product. As part of this I want to unify as much of the build as possible, together with @ybznek. So creating a new product can be as simple as writing a `CMakeLists.txt` file with `ssg_build_product("foobar")`. You will only have to write extra code If you deviate from the standard build procedure.

Rebased to keep up with upstream changes. I need #1389 to proceed.

iankko commented:
@mpreisler 
...
> Rebased to keep up with upstream changes. I need #1389 to proceed.

This is now reviewed && merged.

mpreisler commented:
```
$ mkdir build
$ cd build
$ cmake ../
SCAP Security Guide 0.1.31

Tools:
oscap: /usr/bin/oscap
xsltproc: /usr/bin/xsltproc
xmllint: /usr/bin/xmllint
xmlwf: /usr/bin/xmlwf
 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/mpreisle/d/scap-security-guide/build
$ make fedora
[ 12%] [fedora] generating guide.xml (SVG logo enabled)
[ 25%] [fedora] generating shorthand.xml
[ 37%] [fedora] generating xccdf-unlinked-resolved.xml
Unselected empty groups in 'standard'.
Unselected empty groups in 'common'.
[ 50%] [fedora] generating ocil-unlinked.xml
[ 62%] [fedora] generating xccdf-unlinked-ocilrefs.xml
[ 75%] [fedora] generating bash-remediations.xml

Notification: Merged 35 remediation scripts into XML document.
[ 87%] [fedora] generating xccdf-unlinked-withremediations.xml
[100%] [fedora] generating unlinked-oval.xml (OVAL 5.11 checks enabled)

Notification: Merged 177 OVAL checks into OVAL document.

Notification: this ID is used more than once and should represent equivalent elements: obj_multi_user_wants 
Rewrite the corresponding OVAL checks by placing the identical IDs into its own definition, and extend this definition on necessary places.

Notification: this ID is used more than once and should represent equivalent elements: obj_multi_user_wants 
Rewrite the corresponding OVAL checks by placing the identical IDs into its own definition, and extend this definition on necessary places.

Notification: this ID is used more than once and should represent equivalent elements: obj_multi_user_wants 
Rewrite the corresponding OVAL checks by placing the identical IDs into its own definition, and extend this definition on necessary places.
[100%] Built target fedora
```

mpreisler commented:
I have intentionally broken the build because of file renames. Jenkins will not pass from now on.

mpreisler commented:
As of ee6f2ce this PR can build usable XCCDF 1.1 and OVAL files. Fully linked and resolved.

redhatrises commented:
> As of ee6f2ce this PR can build usable XCCDF 1.1 and OVAL files. Fully linked and resolved.

@mpreisler are you saying that this PR can be reviewed?

mpreisler commented:
> @mpreisler are you saying that this PR can be reviewed?

Not yet, I need to catch up on 14 days of SSG development. Then I need to add datastream and guide building. Then it will be ready.

mpreisler commented:
Hmm, a massive amount of changes and refactoring in the Makefile build system while I was on PTO :-( This will be painful...

mpreisler commented:
@ybznek Could you please help me implement your changes to the cmake build system? It's pretty hard to follow and you will probably get it done way quicker than I would.

I am mainly talking about the remediation scripts.

mpreisler commented:
Hmm, it looks like there is a separate Makefile in the templates directory that generates output from the templates but that output is in the templates directory subtree.

So all in all we run make which runs make which runs make and only then the remediations are built. Plus we have a folder as a Makefile target which is an anti-pattern in most build systems. Could we roll this into `combineremediations.py` or some smart script that will do the right thing and output a **file**?

I think it makes sense to discuss at this point and fix this in SSG first before I attempt to port it to cmake.

mpreisler commented:
After looking at it for a while I managed to fix up the paths. It's really confusing that the output directories are in the **templates** directories! I assumed these directories contains templates and not build output.

Anyway, I got it to work. If you now first run make in the templates dir and then use cmake it builds OVAL and XCCDF successfully.

mpreisler commented:
I now get these legitimate errors that are silenced in the Makefile build system but I won't silence them in the cmake build system:

```
$ make -j4 rhel7
[  6%] [rhel7] generating ssg-rhel7-ds.xml
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-ds.xml' line 36231: Element '{http://checklists.nist.gov/xccdf/1.2}select', attribute 'idref': 'dangling reference to service_telnetd_disabled!' is not a valid value of the atomic type 'xs:NCName'.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-ds.xml' line 36231: Element '{http://checklists.nist.gov/xccdf/1.2}select', attribute 'idref': Warning: No precomputed value available, the value was either invalid or something strange happend.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-ds.xml' line 36233: Element '{http://checklists.nist.gov/xccdf/1.2}select', attribute 'idref': 'dangling reference to uninstall_telnet_server!' is not a valid value of the atomic type 'xs:NCName'.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-ds.xml' line 36233: Element '{http://checklists.nist.gov/xccdf/1.2}select', attribute 'idref': Warning: No precomputed value available, the value was either invalid or something strange happend.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-ds.xml' line 36426: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': 'dangling reference to var_password_max_age!' is not a valid value of the atomic type 'xs:NCName'.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-ds.xml' line 36426: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': Warning: No precomputed value available, the value was either invalid or something strange happend.
OpenSCAP Error: Invalid SCAP Source Datastream (1.2) content in /home/mpreisle/d/scap-security-guide/build/ssg-rhel7-ds.xml. [oscap_source.c:319]
```

These need to be fixed in separate PRs.

mpreisler commented:
RHEL7 ds now builds successfully. RHEL6 has dangling references that need to be fixed:
```
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 28449: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': 'dangling reference to sysctl_net_ipv4_icmp_ignore_bogus_error_messages_value!' is not a valid value of the atomic type 'xs:NCName'.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 28449: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': Warning: No precomputed value available, the value was either invalid or something strange happend.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 28791: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': 'dangling reference to sysctl_net_ipv4_icmp_ignore_bogus_error_messages_value!' is not a valid value of the atomic type 'xs:NCName'.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 28791: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': Warning: No precomputed value available, the value was either invalid or something strange happend.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 29175: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': 'dangling reference to sysctl_net_ipv4_icmp_ignore_bogus_error_messages_value!' is not a valid value of the atomic type 'xs:NCName'.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 29175: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': Warning: No precomputed value available, the value was either invalid or something strange happend.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 29572: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': 'dangling reference to sysctl_net_ipv4_icmp_ignore_bogus_error_messages_value!' is not a valid value of the atomic type 'xs:NCName'.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 29572: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': Warning: No precomputed value available, the value was either invalid or something strange happend.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 29967: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': 'dangling reference to sysctl_net_ipv4_icmp_ignore_bogus_error_messages_value!' is not a valid value of the atomic type 'xs:NCName'.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 29967: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': Warning: No precomputed value available, the value was either invalid or something strange happend.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 30605: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': 'dangling reference to var_password_max_age!' is not a valid value of the atomic type 'xs:NCName'.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 30605: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': Warning: No precomputed value available, the value was either invalid or something strange happend.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 31983: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': 'dangling reference to sysctl_net_ipv4_icmp_ignore_bogus_error_messages_value!' is not a valid value of the atomic type 'xs:NCName'.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml' line 31983: Element '{http://checklists.nist.gov/xccdf/1.2}refine-value', attribute 'idref': Warning: No precomputed value available, the value was either invalid or something strange happend.
OpenSCAP Error: Invalid SCAP Source Datastream (1.2) content in /home/mpreisle/d/scap-security-guide/build/ssg-rhel6-ds.xml. [oscap_source.c:319]
RHEL/6/CMakeFiles/rhel6.dir/build.make:84: recipe for target 'ssg-rhel6-ds.xml' failed
make[2]: *** [ssg-rhel6-ds.xml] Error 1
make[2]: *** Deleting file 'ssg-rhel6-ds.xml'
CMakeFiles/Makefile2:140: recipe for target 'RHEL/6/CMakeFiles/rhel6.dir/all' failed
make[1]: *** [RHEL/6/CMakeFiles/rhel6.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
```

mpreisler commented:
RHEL5 invalid refs:
```
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel5-ds.xml' line 34663: Element '{http://checklists.nist.gov/xccdf/1.2}select', attribute 'idref': 'dangling reference to file_groupowner_etc_shadow!' is not a valid value of the atomic type 'xs:NCName'.
File '/home/mpreisle/d/scap-security-guide/build/ssg-rhel5-ds.xml' line 34663: Element '{http://checklists.nist.gov/xccdf/1.2}select', attribute 'idref': Warning: No precomputed value available, the value was either invalid or something strange happend.
OpenSCAP Error: Invalid SCAP Source Datastream (1.2) content in /home/mpreisle/d/scap-security-guide/build/ssg-rhel5-ds.xml. [oscap_source.c:319]
RHEL/5/CMakeFiles/rhel5.dir/build.make:84: recipe for target 'ssg-rhel5-ds.xml' failed
```

mpreisler commented:
I have fixed the dangling reference in RHEL5 in 9882b7a7402f4f18f6def58d571b19a08ea7555a, now it builds using the cmake build system:
```
[  6%] [rhel5] generating shorthand.xml
[ 13%] [rhel5] generating xccdf-unlinked-resolved.xml
[ 20%] [rhel5] generating ocil-unlinked.xml
[ 26%] [rhel5] generating xccdf-unlinked-ocilrefs.xml
[ 33%] [rhel5] generating xccdf-unlinked.xml
[ 40%] [rhel5] linking IDs in XCCDF, OVAL and OCIL files
[ 46%] [rhel5] generating ssg-rhel5-xccdf.xml
[ 53%] [rhel5] generating ssg-rhel5-ocil.xml
[ 60%] [rhel5] generating ssg-rhel5-oval.xml
Unselected empty groups in 'stig-rhel5-server-upstream'.
[ 66%] [rhel5] generating ssg-rhel5-xccdf-1.2.xml
[ 73%] [rhel5] generating ssg-rhel5-ds.xml
[100%] Built target rhel5
```

mpreisler commented:
I have fixed up the rest of the RHEL7 danging references. The cmake build system now builds RHEL5, 6, 7 and Fedora successfully, including HTML guides.

```
$ time make -j 4
...
[ 98%] [rhel5] generating HTML guides for all profiles in ssg-rhel5-ds.xml
[100%] [rhel7] generating HTML guides for all profiles in ssg-rhel7-ds.xml
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel5-guide-default.html' for profile ID '' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-5'.
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel5-guide-stig-rhel5-server-upstream.html' for profile ID 'xccdf_org.ssgproject.content_profile_stig-rhel5-server-upstream' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-5'.
[100%] Built target rhel5
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-guide-common.html' for profile ID 'xccdf_org.ssgproject.content_profile_common' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-7'.
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-guide-cjis-rhel7-server.html' for profile ID 'xccdf_org.ssgproject.content_profile_cjis-rhel7-server' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-7'.
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-guide-C2S.html' for profile ID 'xccdf_org.ssgproject.content_profile_C2S' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-7'.
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-guide-nist-cl-il-al.html' for profile ID 'xccdf_org.ssgproject.content_profile_nist-cl-il-al' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-7'.
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-guide-pci-dss.html' for profile ID 'xccdf_org.ssgproject.content_profile_pci-dss' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-7'.
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-guide-rht-ccp.html' for profile ID 'xccdf_org.ssgproject.content_profile_rht-ccp' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-7'.
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-guide-stig-rhel7-server-upstream.html' for profile ID 'xccdf_org.ssgproject.content_profile_stig-rhel7-server-upstream' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-7'.
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-guide-stig-rhel7-server-gui-upstream.html' for profile ID 'xccdf_org.ssgproject.content_profile_stig-rhel7-server-gui-upstream' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-7'.
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-guide-docker-host.html' for profile ID 'xccdf_org.ssgproject.content_profile_docker-host' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-7'.
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-guide-standard.html' for profile ID 'xccdf_org.ssgproject.content_profile_standard' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-7'.
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-guide-stig-rhel7-workstation-upstream.html' for profile ID 'xccdf_org.ssgproject.content_profile_stig-rhel7-workstation-upstream' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-7'.
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-guide-ospp-rhel7-server.html' for profile ID 'xccdf_org.ssgproject.content_profile_ospp-rhel7-server' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-7'.
Generated '/home/mpreisle/d/scap-security-guide/build/ssg-rhel7-guide-default.html' for profile ID '' in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-7'.
[100%] Built target rhel7

real	0m12.066s
user	0m38.076s
sys	0m2.489s
```

mpreisler commented:
OK, so now the cmake build system can build all the files and validate them. I think it's time other SSG community members gave this a try. Feedback greatly appreciated!

How to get started:
```
cd scap-security-guide
cd cmake
# this is a hack for now, building remediations from templates isn't in cmake yet
./build_remediations.sh
cd ..
mkdir build
cd build
cmake ../
# build everything
make -j4
# builds all products
make all -j4
# just the product
make rhel7 -j4
# validate validates all
make validate -j4
# just the product
make rhel7-validate -j4
# just one file of a product
make rhel7-validate-xccdf -j4 
# clean the build folder
make clean
```

The SCAP file builds are fully incremental with proper deps. I want to make the validate targets the same, should have that in a few hours.

Lemme know what you think, @iankko @shawndwells @redhatrises @ybznek @jan-cerny

mpreisler commented:
TODO:
- incremental validation, only validate if the file changed
- port all products to cmake
- build remediations from templates using cmake
- verify-references.py

mpreisler commented:
You can build with `ninja-build` instead of `make`. The biggest advantage is that outputs of the jobs don't get mixed up when running multiple jobs in parallel. To use Ninja:

```
cd build
cmake -G "Ninja" ..
ninja-build -j 4
ninja-build -j 4 validate
...
```

ybznek commented:
@mpreisler Here is python script to build remediations.
https://github.com/OpenSCAP/scap-security-guide/compare/master...ybznek:master-build-remediations.py

```
cd <ssg root dir>
./build_remediations.py 
./build_remediations.py clean
```

I can modify/improve it. But I need to have defined requirements.

mpreisler commented:
Now also building derivatives, with the option to turn them on or off in the configuration step.

mpreisler commented:
@ybznek 

> I can modify/improve it. But I need to have defined requirements.

I also need a way to get all inputs that are to be used and all outputs that are to be written out. That way I can create the cmake targets. Other than that it looks good to me, could you please submit it as a PR?

mpreisler commented:
> Lemme know what you think, @iankko @shawndwells @redhatrises @ybznek @jan-cerny

Didn't hear from anyone in 11 days so I am totally gonna spam you again :-) Sorry not sorry.

radzy commented:
@mpreisler : I'd be happier reviewing this if the BUILD.md file were updated.  I believe this is far enough along that the new contents wouldn't be too likely to change significantly.

And if you'd like to add some notes on a newbie guide of some sort, indicating what needs to be done to bootstrap a new platform, I'd be happy to work with that, verify it, and add some of the notes I've taken.  NEW_PLATFORM.md ?  Again, I don't think this information would be likely to change too much before this change is approved.

redhatrises commented:
@mpreisler so far so good. I am really liking this port! The sooner we can finish this port the better.

One small request, can we just go a head and add the `build` directory with a `.gitignore` or `.gitkeep` inside of it. I know the build directory is arbitrary, but it can be useful in the sense that it is one less step not to have to create a build directory.

mpreisler commented:
@redhatrises Thx, done in f45c878bd3c91ddd9d90e5943c3a07797d7bbe79

mpreisler commented:
> @mpreisler : I'd be happier reviewing this if the BUILD.md file were updated. I believe this is far enough along that the new contents wouldn't be too likely to change significantly.

Done in 2576fa8

> And if you'd like to add some notes on a newbie guide of some sort, indicating what needs to be done to bootstrap a new platform, I'd be happy to work with that, verify it, and add some of the notes I've taken. NEW_PLATFORM.md ? Again, I don't think this information would be likely to change too much before this change is approved.

This is still going to change so let's not work on that here. One big goal of the cmake rewrite is to have no copy pasted code in the build system. Adding a new platform is just adding one file with one line in the simple cases. I have yet to solve the issue of creating the directory structure for a new product.

redhatrises commented:
@mpreisler is this TODO list still valid?

TODO:

 - ~~incremental validation, only validate if the file changed~~
- generate CPE dictionaries
- port all products to cmake
- generate HTML tables
- build remediations from templates using cmake
- ~~verify-references.py~~
- ~~build derivatives~~
- PCI centric benchmark
- multibenchmark SDS
- install target

Just curious where we are with this PR.

mpreisler commented:
@redhatrises Yes, the TODO is still valid. I keep rebasing this but I had other high priority tasks so I couldn't touch this for a while.

Volunteers to help me with this would be greatly appreciated!

redhatrises commented:
> Volunteers to help me with this would be greatly appreciated!

@mpreisler how would you like volunteers to help you? Submit PRs to your PR branch?

mpreisler commented:
@redhatrises Either that or if you do have SSG push rights you can already just push to my branch.

redhatrises commented:
@mpreisler what do you think of 6b1ff57f3a563957e671b34ff56517baa1894c8e for building remediations from templates?

mpreisler commented:
Looks good.

Please don't add merge commits and commits already in master. This will make the branch messy. Do `git rebase -i master` to keep it clean. That's really the only way to manage branches long lived branches. I will fix it tomorrow.

mpreisler commented:
Rebased to keep things clean.

mpreisler commented:
ping @redhatrises you may need to `git pull --rebase` your local cmake branch against the remote sooner rather than later to avoid conflicts.

redhatrises commented:
> ping @redhatrises you may need to git pull --rebase your local cmake branch against the remote sooner rather than later to avoid conflicts.

Already done.

mpreisler commented on cmake/SSGCommon.cmake:
```diff
@@ -167,7 +168,8 @@ macro(ssg_build_oval_unlinked PRODUCT)
             OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/oval-unlinked.xml
             # TODO: config
-            COMMAND RUNTIME_OVAL_VERSION=5.11 ${SSG_SHARED_UTILS}/combine-ovals.py ${CMAKE_SOURCE_DIR}/config ${PRODUCT} ${OVAL_DEPS_DIR} ${OVAL_511_DEPS_DIR} ${SHARED_OVAL_DEPS_DIR} ${SHARED_OVAL_511_DEPS_DIR} > ${CMAKE_CURRENT_BINARY_DIR}/oval-unlinked.xml
+            COMMAND SHARED=${SSG_SHARED} ${SSG_SHARED_UTILS}/generate-from-templates.py --oval_version ${OSCAP_OVAL_VERSION} --input ${CMAKE_CURRENT_SOURCE_DIR}/templates --output ${CMAKE_CURRENT_BINARY_DIR}/build --language oval build
```
CMAKE_CURRENT_BINARY_DIR/build ? are you sure?

mpreisler replied:
I'd build remediations into CURRENT_BINARY_DIR/remediations/$language and OVAL into CURRENT_BINARY_DIR/oval

CURRENT_BINARY_DIR/build looks like a typo. I would end up as `git_root/build/RHEL/7/build`

mpreisler commented on CMakeLists.txt:
```diff
@@ -149,6 +149,23 @@ install(FILES "${CMAKE_SOURCE_DIR}/Contributors.md"
+    COMMAND rm -f ${CMAKE_BINARY_DIR}/*.xml
+    COMMAND find ${CMAKE_BINARY_DIR}/ -type d -name oval | xargs rm -rf
+    COMMAND find ${CMAKE_BINARY_DIR}/ -type d -name remediations | xargs rm -rf
+)
```
Please remove clean-all target. Clean is already provided by cmake.

redhatrises replied:
@mpreisler cmake doesn't clean up all the `html` files or the `oval/remediations` directories. Should we account for that at all?

mpreisler replied:
That's a bug in the cmake targets then, we have side-effects not listed in cmake. Let's not do workarounds to hide bugs pls.

mpreisler commented:
@redhatrises what do you think about moving all the remediation building to `ssg_build_remediations`? I don't think it will change behavior but will make things cleaner. Right now we build some remediations there and the rest in `build_xccdf_with_remediations`. The latter macro was initially supposed to take built remediations and put them in the XCCDF.

redhatrises commented:
> @redhatrises what do you think about moving all the remediation building to `ssg_build_remediations`?

@mpreisler sure. Unless there is a reason for a second macro besides it being easier to read especially if more remediation scripts are added.

mpreisler commented:
> @mpreisler sure. Unless there is a reason for a second macro besides it being easier to read especially if more remediation scripts are added.

I don't think there is a reason to introduce more macros but we may split the targets up to get more parallelism. We can do that later though and I think the performance to be gained from that will likely be minimal.

redhatrises commented:
@mpreisler what all needs to be done with `make install` seems like it is already done? Or am I missing something?

mpreisler commented:
> @mpreisler what all needs to be done with make install seems like it is already done? Or am I missing something?

There is a bunch of things that it doesn't install. Kickstarts, HTML tables, stuff like that. I wouldn't cross it off before the diff between `make install` of the old build system and cmake are the same.

mpreisler commented:
@redhatrises We need to change the add_to_sds macros, right now they have ssg-...-ds.xml as both input as output. That means that they are always outdated and always need to be rebuilt. Other than that though, awesome that you added support for PCI-DSS and OCIL.

redhatrises commented:
> @redhatrises We need to change the add_to_sds macros, right now they have ssg-...-ds.xml as both input as output. That means that they are always outdated and always need to be rebuilt. Other than that though, awesome that you added support for PCI-DSS and OCIL.

@mpreisler recommendations for what the files should be as output?

mpreisler commented:
Rebased, fixed build.

Fedora CPE build is broken, @redhatrises could you help me with that? Is Fedora CPE even supposed to work?

This is very close to being merge-able.

mpreisler commented:
> Fedora CPE build is broken, @redhatrises could you help me with that? Is Fedora CPE even supposed to work?

I have added a dummy CPE dict to fix the situation for the time being.

mpreisler commented:
A lot of the cmake targets are overriding the same targets or having outputs that don't match what they actually output. I am not sure when this happened :-/ Will have to fix this before we move forward.

redhatrises commented:
> Fedora CPE build is broken, @redhatrises could you help me with that? Is Fedora CPE even supposed to work?

@mpreisler I can fix the Fedora CPE for you. Will get that tomorrow.

redhatrises commented:
@mpreisler will also be fixing ssg_build_add_pci_dss_to_sds tomorrow as well.

mpreisler commented:
> @mpreisler will also be fixing ssg_build_add_pci_dss_to_sds tomorrow as well.

I just fixed that, we are back to "perfect" builds. If there are no changes it will not rebuild anything.

mpreisler commented:
cmake was reporting a circular dependency around the HTML guides but I can't reproduce that after wiping the build folder. Can anyone reproduce that?

mpreisler commented:
> @mpreisler I can fix the Fedora CPE for you. Will get that tomorrow.

Awesome, I put in a dummy file just to get the build to go through but it turns out it doesn't help much. I think we need the real CPE dict file.

redhatrises commented:
> cmake was reporting a circular dependency around the HTML guides but I can't reproduce that after wiping the build folder. Can anyone reproduce that?

I actually have, but I don't remember why. I will see if I have the same issue as well.

mpreisler commented:
Updated TODO:

 - ~~incremental validation, only validate if the file changed~~
- ~~generate CPE dictionaries~~
- ~~port all products to cmake~~
- generate HTML tables
- ~~build remediations from templates using cmake~~
- ~~verify-references.py~~
- ~~build derivatives~~
- ~~PCI centric benchmark~~
- ~~multibenchmark SDS~~
- ~~install target~~

We need to wrap up the HTML tables and then I think this is ready for a thorough review and perhaps even merging.

redhatrises commented:
> We need to wrap up the HTML tables and then I think this is ready for a thorough review and perhaps even merging.

@mpreisler big thing there is how to handle the different profiles with the different products.

redhatrises commented:
@mpreisler added Fedora CPE dict which looks like we get complete builds. Now on to the HTML tables.

redhatrises commented:
@mpreisler this may not be ideal, but what about having the profiles in the the local `cmakelist.txt` to the product? Then, `cmake/SSGCommon.cmake` generates profiles for however many table profiles are in `cmakelist.txt`.

mpreisler commented:
> @mpreisler this may not be ideal, but what about having the profiles in the the local cmakelist.txt to the product? Then, cmake/SSGCommon.cmake generates profiles for however many table profiles are in cmakelist.txt.

Yeah, I was thinking about some macro and we can call it manually in the per product CMakeLists. It doesn't even have to be in the shared product build code. I am writing it as we speak. Will add a few and then we can collaborate to get them all in.

mpreisler commented:
@redhatrises There already are targets for various HTML tables in cmake but they are never used. Did you just add them blind or did you make it work in some way? Either we broke it or it never worked in the first place.

mpreisler commented:
I will redo the HTML tables, they are implemented as if every product had the same XSLTs and the same profiles.

redhatrises commented:
> @redhatrises There already are targets for various HTML tables in cmake but they are never used. Did you just add them blind or did you make it work in some way? Either we broke it or it never worked in the first place.

I just added them. I was going to go back and fix them later when the profile handling was figured out.

redhatrises commented:
@mpreisler anything I can help you with on HTML tables stuff?

mpreisler commented:
> @mpreisler anything I can help you with on HTML tables stuff?

If you want to take a look at this you could use the new macros to build RHEL6 HTML tables.

mpreisler commented:
[rhel7.svg.zip](https://github.com/OpenSCAP/scap-security-guide/files/769683/rhel7.svg.zip)

Attaching RHEL7 build graph SVG. It is pretty wild but it seems correct to me. Would appreciate a review.

mpreisler commented:
Wrapped the custom commands in proper cmake targets. That means that dependencies can now be shared. In other words, this now works properly:

```
cmake ../ && make rhel7-guides rhel7-validate -j
```

And it takes less than a second for the build system to realize it doesn't need to regenerate anything if you run it twice.

mpreisler commented:
@redhatrises I think I got the RHEL7 tables down, including installing them. The stig-testinfo.html is empty but the current old build system has the same issue, so that's unrelated to cmake.

Could you take a look at it and possibly implement RHEL6 table building using the macros I created?

After we do that I think this is ready for merging.

mpreisler commented:
TODO:
- remove input and output folders, do proper file deps - this should be easy because our build scripts have options to output inputs and outputs
- implementing something that figures out deps of the guide.xml target. We need to go into the XSLT and recursively gather all included files.

mpreisler commented:
The way we build OVAL from templates is very suspect. We do not build from the shared templates. Will have to take a look at that before I introduced proper deps and outputs for OVALs.

mpreisler commented:
Here is a little demo in case you are too lazy to test this out: https://www.youtube.com/watch?v=PFsuTrxX9Lo :-) In the video I use ninja-build because that's the fastest way to build, the same thing works with the conventional GNU `make` build system.

mpreisler commented:
@redhatrises The circular deps was caused by the HTML guides glob. We were gathering inputs and outputs at configure time. That can't work because at configure time we don't have the source datastream. I have fixed it by removing the blog.

Ofc that means make clean won't remove those HTMLs but that's a less serious issue than having circular deps.

mpreisler commented:
@ybznek 

```
 $ ../shared/utils/generate-from-templates.py --shared ../shared/ --oval_version oval_5.11 --input ../RHEL/7/templates/ --output ./RHEL/7/ list-inputs
/home/mpreisle/d/scap-security-guide/shared/templates/template_BASH_accounts_password
/home/mpreisle/d/scap-security-guide/shared/templates/template_ANSIBLE_permissions
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/csv/sysctl_values.csv
/home/mpreisle/d/scap-security-guide/shared/templates/template_ANACONDA_package_removed
/home/mpreisle/d/scap-security-guide/shared/templates/template_BASH_kernel_module_disabled
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/oval_5.11_templates/template_package_removed
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/csv/services_enabled.csv
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/template_sysctl_runtime_var
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/template_sysctl_runtime
/home/mpreisle/d/scap-security-guide/shared/templates/template_BASH_permissions
/home/mpreisle/d/scap-security-guide/shared/templates/template_ANSIBLE_kernel_module_disabled
/home/mpreisle/d/scap-security-guide/shared/templates/create_permission.py
/home/mpreisle/d/scap-security-guide/shared/templates/template_BASH_sysctl
/home/mpreisle/d/scap-security-guide/shared/templates/template_PUPPET_package_installed
/home/mpreisle/d/scap-security-guide/shared/templates/create_package_removed.py
/home/mpreisle/d/scap-security-guide/shared/templates/template_ANSIBLE_package_removed
/home/mpreisle/d/scap-security-guide/shared/templates/template_BASH_package_installed
/home/mpreisle/d/scap-security-guide/shared/templates/template_ANACONDA_package_installed
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/template_BASH_service_disabled
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/csv/services_disabled.csv
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/template_sysctl_ipv6
/home/mpreisle/d/scap-security-guide/shared/templates/template_PUPPET_service_disabled
/home/mpreisle/d/scap-security-guide/shared/templates/template_BASH_sysctl_var
/home/mpreisle/d/scap-security-guide/shared/templates/create_services_disabled.py
/home/mpreisle/d/scap-security-guide/shared/templates/template_BASH_service_enabled
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/template_sysctl_static
/home/mpreisle/d/scap-security-guide/shared/templates/create_services_enabled.py
/home/mpreisle/d/scap-security-guide/shared/templates/create_sysctl.py
/home/mpreisle/d/scap-security-guide/shared/templates/template_ANSIBLE_service_disabled
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/csv/oval_5.11/packages_removed.csv
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/csv/file_dir_permissions.csv
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/csv/packages_removed.csv
/home/mpreisle/d/scap-security-guide/shared/templates/template_ANSIBLE_service_enabled
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/csv/accounts_password.csv
/home/mpreisle/d/scap-security-guide/shared/templates/template_ANSIBLE_package_installed
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/csv/kernel_modules_disabled.csv
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/csv/packages_installed.csv
/home/mpreisle/d/scap-security-guide/shared/templates/template_PUPPET_service_enabled
/home/mpreisle/d/scap-security-guide/shared/templates/template_PUPPET_package_removed
/home/mpreisle/d/scap-security-guide/shared/templates/create_package_installed.py
/home/mpreisle/d/scap-security-guide/shared/templates/create_kernel_modules_disabled.py
/home/mpreisle/d/scap-security-guide/shared/templates/template_BASH_package_removed
/home/mpreisle/d/scap-security-guide/shared/templates/create_accounts_password.py
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/template_kernel_module_disabled
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/template_sysctl
/home/mpreisle/d/scap-security-guide/RHEL/7/templates/template_sysctl_static_var
```

Is this expected to contain shared templates? I thought we were supposed to run genreate-from-templates twice, once for per product OVALs and once for shared OVALs. That's what we do for remediations. Has this changed after I implemented it in cmake? Should be only run it once now? How will we separate shared remediations and shared OVALs in the future when they are built as part of the normal per product process?

ybznek commented:
@mpreisler 
We should run it twice.
But If you want to generate something from csv in product and you don't have template for it, it will use shared template.
- User probably wants to have same permission check for `/etc/hosts` - we can share generated files across all distros
- But in one product user could want to check/remediate permission of `/<mySpecialDistroFileNotExistingInAnotherDistro>` and in another product user wants check `/<myAnotherDistroSpecialFile>`. It doesn't make sense to have remediation_template implemented twice in both products. So it will use shared template.

- I don't think that it is completely different behavior than before. In previous version were probably duplicates between product/shared template, so it wasn't using shared template_* in product.

It should not be complicated to find all files, which are generated in product, and use shared template. Such files we can start to generate in shared directory.

Currently I am still removing generated files. We can make cleaner structure in
https://github.com/OpenSCAP/scap-security-guide/issues/1668

mpreisler commented:
Had a meeting with @ybznek to figure out how to build OVALs in cmake. We figured it out and I implemented it in this branch.

Still need to implement tarball and rpm (aaarrrrghhhh).

mpreisler commented:
`make rpm` works
`make tarball` works

Please review and provide feedback!

mpreisler commented:
Note: Obviously jenkins will fail because of how different the build system is. After merging we will have to change Jenkins config.

shawndwells commented on BUILD.md:
```diff
@@ -2,61 +2,90 @@
+1. On Red Hat Enterprise Linux and Fedora make sure the packages `openscap-utils`, `openscap-python`, and `python-lxml` and their dependencies are installed. We require version `1.0.8` or later of `openscap-utils` (available in Red Hat Enterprise Linux) as well as `git`.
 
- `# yum -y install git openscap-utils openscap-python python-lxml`
+ `# yum -y install git cmake openscap-utils openscap-python python-lxml`
```
Add a reference to cmake in the prose, e.g. "..... make sure the packages **cmake**, openscap-utils, openscap-python...."

mpreisler replied:
Good catch, fixed.

shawndwells commented on BUILD.md:
```diff
@@ -2,61 +2,90 @@
 
  `# dnf -y install ShellCheck`
 
-2. Download the source code 
```
Should an optional note about installing Ninja be included here?

mpreisler replied:
Done

mpreisler commented:
@openscap-jenkins skip ci

shawndwells commented:
had a chance to test the cmake build -- holy crap this is fast! Building RHEL7 took 0.491 seconds!

Getting errors on the overall build - is this still expected?
``
[ 67%] [firefox] generating ssg-firefox-ds.xml
File '/var/www/html/scap-security-guide/build/ssg-fedora-ds.xml' line 34: Element '{http://oval.mitre.org/XMLSchema/oval-definitions-5}extend_definition': No match found for key-sequence ['oval:ssg-package_audit_installed:def:1'] of keyref '{http://oval.mitre.org/XMLSchema/oval-definitions-5}extendKeyRef'.
File '/var/www/html/scap-security-guide/build/ssg-fedora-ds.xml' line 50: Element '{http://oval.mitre.org/XMLSchema/oval-definitions-5}extend_definition': No match found for key-sequence ['oval:ssg-package_chrony_installed:def:1'] of keyref '{http://oval.mitre.org/XMLSchema/oval-definitions-5}extendKeyRef'.
File '/var/www/html/scap-security-guide/build/ssg-fedora-ds.xml' line 66: Element '{http://oval.mitre.org/XMLSchema/oval-definitions-5}extend_definition': No match found for key-sequence ['oval:ssg-package_cronie_installed:def:1'] of keyref '{http://oval.mitre.org/XMLSchema/oval-definitions-5}extendKeyRef'.
File '/var/www/html/scap-security-guide/build/ssg-fedora-ds.xml' line 82: Element '{http://oval.mitre.org/XMLSchema/oval-definitions-5}extend_definition': No match found for key-sequence ['oval:ssg-package_firewalld_installed:def:1'] of keyref '{http://oval.mitre.org/XMLSchema/oval-definitions-5}extendKeyRef'.
OpenSCAP Error: Invalid SCAP Source Datastream (1.2) content in /var/www/html/scap-security-guide/build/ssg-fedora-ds.xml. [oscap_source.c:319]
make[2]: *** [ssg-fedora-ds.xml] Error 1
make[1]: *** [Fedora/CMakeFiles/generate-ssg-fedora-ds.xml.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
No extended-components, nothing to do...
[ 67%] Built target generate-ssg-firefox-ds.xml
[ 67%] Built target generate-ssg-rhel6-xccdf-1.2.xml
[ 67%] Built target generate-ssg-rhel7-xccdf-1.2.xml
make: *** [all] Error 2

``

shawndwells commented:
^^ that's from ``cmake ../ ; make -j4``

redhatrises commented:
@shawndwells the errors you are getting are not related to Cmake but to the build templates.

redhatrises commented:
@shawndwells looks like the remediations/ovals that are generated on build are not found/merged into the built products.

mpreisler commented:
Seems like we need changes in #1715 for the Fedora build to work with cmake. I wonder why it's not needed for the old style build systems, they both use the same commands :-/

redhatrises commented:
@mpreisler it also looks like `RHEL/7/input/oval/oval_5.11` is not being included in the build which is an issue. It looks like it is included in the `OVAL_511_COMBINE_PATHS`, so it should pull in those OVAL files, no?

mpreisler commented:
> @mpreisler it also looks like RHEL/7/input/oval/oval_5.11 is not being included in the build which is an issue. It looks like it is included in the OVAL_511_COMBINE_PATHS, so it should pull in those OVAL files, no?

I would think so but I will double check.

mpreisler commented:
> Getting errors on the overall build - is this still expected?

This is now fixed. #1715 has been merged and I rebased the cmake branch.

mpreisler commented:
> @mpreisler it also looks like RHEL/7/input/oval/oval_5.11 is not being included in the build which is an issue. It looks like it is included in the OVAL_511_COMBINE_PATHS, so it should pull in those OVAL files, no?
> I would think so but I will double check.

There was a typo in the variable, I must have made it when refactoring. Will fix this shortly.

mpreisler commented:
@shawndwells @redhatrises I have fixed the OVAL 5.11 issues and made some other minor improvements to enabling OVAL 5.11 at configure time.

Please re-review.

redhatrises commented:
@mpreisler it still doesn't seem to be pulling in the content from `RHEL/7/input/oval/oval_5.11`. Is there a `combine-ovals.py` issue not getting the path?

mpreisler commented:
> @mpreisler it still doesn't seem to be pulling in the content from RHEL/7/input/oval/oval_5.11. Is there a combine-ovals.py issue not getting the path?

Make sure you are using the updated branch. I have rebased and push-forced.

redhatrises commented:
>> @mpreisler it still doesn't seem to be pulling in the content from RHEL/7/input/oval/oval_5.11. Is there a combine-ovals.py issue not getting the path?
>
>Make sure you are using the updated branch. I have rebased and push-forced.

@mpreisler I am.

mpreisler commented:
@redhatrises OK, I have made another bug while improving things :-D I will stop improving things after I fix it.

mpreisler commented:
> @mpreisler I am.

@redhatrises Fixed now.

mpreisler commented:
I found another issue, OVAL 5.10 builds fine now but OVAL 5.11 has nonexistent refs. Will fix.

mpreisler commented:
The `generate-from-templates.py` script mysteriously stopped generating OVAL 5.11... What the f***... If anybody can see why I would appreciate it.

mpreisler commented:
ninja-build has a few debugging options that might help with review. Check out:

```
# list all commands required to build rhel7
ninja-build -t commands rhel7

# browse the dependency graph and targets
ninja-build -t browse
```

redhatrises commented:
@mpreisler a couple of things before I think that we can merge:

- `xmlns:xhtml="http://www.w3.org/1999/xhtml"` is added to every `<fix />` element for every final product XCCDF file which is not currently done with the current build system. Is this an error in the new or old build system?
- `<ns0:sub xmlns:ns0="http://checklists.nist.gov/xccdf/1.1" idref="function_package_command"/>` is created instead of `<sub idref="function_package_command"/>` for every final product XCCDF file which is not currently done with the current build system. Is this an error in the new or old build system?
- `<select idref="srg_support" selected="false">` is missing from being added in each profile which is not currently not an issue with the current build system.

rsprudencio commented:
I tested it here, it built everything and I tried to scan with Fedora datastream and it worked as expected.

Great work :smile:

mpreisler commented:
> xmlns:xhtml="http://www.w3.org/1999/xhtml" is added to every <fix /> element for every final product XCCDF file which is not currently done with the current build system. Is this an error in the new or old build system?
> <ns0:sub xmlns:ns0="http://checklists.nist.gov/xccdf/1.1" idref="function_package_command"/> is created instead of <sub idref="function_package_command"/> for every final product XCCDF file which is not currently done with the current build system. Is this an error in the new or old build system?
> <select idref="srg_support" selected="false"> is missing from being added in each profile which is not currently an issue with the current build system.

The first two are not affecting anything in the real world. I will take a look at it but it is probably added by oscap. I would prefer not to do any seds or workarounds, the namespaces don't do any harm anywhere.

I will fix the srg_support unselect.

mpreisler commented:
> I tested it here, it built everything and I tried to scan with Fedora datastream and it worked as expected.
> 
> Great work

Did you take a look at the cmake sources as well or just functional testing?

redhatrises commented:
> @redhatrises 
>
> The first two are not affecting anything in the real world. I will take a look at it but it is probably added by oscap. I would prefer not to do any seds or workarounds, the namespaces don't do any harm anywhere.

It just seems a little odd that it is in the XCCDF but is stripped from the DS files.

>I will fix the srg_support unselect.

I say once that is done that we merge this PR and enhance/fix what other issue/request later. This builds and the XCCDF/DS is mostly similar (differences are not an issue) to the old build system.

mpreisler commented:
@redhatrises I think I fixed it. xccdf-linked.xml has the srg_support Group and it is removed in the final ssg-$PRODUCT-xccdf.xml

The SRG maps are generated from the xccdf-linked.xml, other tables are generated from the final XCCDF.

The unselect is not necessary because the Groups don't even appear in the final XCCDF. In the old build system this is most likely a bug.

mpreisler commented:
@redhatrises if you give me 10 mins I will fix the unnecessary xmlns as well.

mpreisler commented:
@redhatrises Fixed all three. Please review.

redhatrises commented:
I say ACK. Unless there is more that we need to test/look at. @mpreisler if you don't think that anything else needs to be done, how about you do the merging honors of this awesome PR!

mpreisler commented:
@redhatrises thx for your vote of confidence. Since this changes how everything work let's give people one more day to voice their opinion. If nobody speaks up by tomorrow noon EST I will merge it.

mpreisler commented:
I will fix up Jenkins shortly

ybznek commented:
Fedora/WRLinux and oval5.11: ```<definition class="inventory" id="oval:ssg-installed_OS_is_centos7:def:1" version="2">```
	Such types of oval definitions are twice in DS. It will need some investigation.
